extensions/amp-bind/0.1/amp-state.js

```diff
@@ -207,15 +210,35 @@ export class AmpState extends AMP.BaseElement {
   /**
    * @param {boolean} isInit
    * @param {boolean=} opt_refresh
-   * @return {!Promise<undefined>}
+   * @return {!Promise}
    * @private
    */
   fetchAndUpdate_(isInit, opt_refresh) {
     // Don't fetch in prerender mode.
-    return this.getAmpDoc()
+    const fetchAndUpdatePromise = this.getAmpDoc()
       .whenFirstVisible()
       .then(() => this.prepareAndSendFetch_(isInit, opt_refresh))
       .then(json => this.updateState_(json, isInit));
+
+    if (isInit) {
+      Services.bindForDocOrNull(this.element).then(bind => {
+        devAssert(bind);
+        bind.registerAsyncAmpState(
+          this.element.getAttribute('id'),
+          fetchAndUpdatePromise
+        );
+      });
```

Comment:
Would this be cleaner inverted? I.e. if the `Bind` service called each `AmpState` instead. Similar to what happens in `Bind.initialize_()` now calling `parseAndUpdate()`.

Then you wouldn't need to maintain a map of id-to-Promise, and you wouldn't have to worry about the order of `AmpState.buildCallback` vs. `Bind.rescan`.

Reply:
I hadn't considered the `buildCallback` vs. `rescan` race! I'm probably not properly handling that right now.

And hmm. Waiting for all the `srcs` to resolve within `Bind.initialize_()` is definitely the simplest option, but then could unnecessarily slow down `eval` for cases that don't depend on async data (e.g. for pages with with some `src` loaded data and some inline json)

Reply:
A second place where `Bind.initialize_` differs is in cases where someone is attempting to sequence an `amp-state` refresh and a `Bind.setState`. For example, `on="tap:ampState.refresh(), AMP.setState({ thing: ampState})"`.

I'm not sure either of these points matter, so I'll opt for the simpler impl you suggest 👍

Reply:
I'm not suggesting that we chain `Bind.initialize_` to the fetch promises; it's just an example. We should do it in `getStateWithWait`. Can share some code.

Reply:
I think I get it. Do a dom scan for `amp-state` elements within the `getStateWithWait` fn. The ivar map can just be a local var then. I'll update the core logic here and lmk if it is along the lines of what you were thinking.